compiler/noirc_evaluator/src/ssa/opt/preprocess_fns.rs

```diff
@@ -0,0 +1,37 @@
+//! Pre-process functions before inlining them into others.
+
+use crate::ssa::Ssa;
+
+use super::inlining;
+
+impl Ssa {
+    /// Run pre-processing steps on functions in isolation.
+    pub(crate) fn preprocess_functions(
+        mut self,
+        aggressiveness: i64,
+        max_bytecode_increase_percent: Option<i32>,
+    ) -> Ssa {
+        // Bottom-up order, starting with the "leaf" functions, so we inline already optimized code into the ones that call them.
+        let bottom_up = inlining::compute_bottom_up_order(&self);
+        let not_to_inline = inlining::get_functions_to_inline_into(&self, false, aggressiveness);
+
+        for id in bottom_up.into_iter().filter(|id| !not_to_inline.contains(id)) {
```

Comment:
@TomAFrench do you reckon that this kind of bottoms-up ordering and the folding functions makes sense? Maybe not exactly like this (you mentioned doing stuff in parallel, and limiting which function it gets applied to), but as a general direction?

Reply:
Yeah, it looks sensible to me. The execution failures we're getting on the protocol circuits seem real though (rather than just having stale inputs) so something is up.

Reply:
It was because of calling the DIE pass; included an example in the PR description:

```
After Removing Paired rc_inc & rc_decs:
...
acir(inline) fn Add10 f1 {
  b0(v0: &mut Field):
    v1 = load v0 -> Field
    v2 = load v0 -> Field
    v4 = add v2, Field 10
    store v4 at v0
    return
}

After Pre-processing Functions:
...
acir(inline) fn Add10 f1 {
  b0(v0: &mut Field):
    return                              # Whole function body got eliminated!
}
```

It sees `store v4 at v0` as a dead instruction.

Reply:
Looks like we need to pass some extra context to the `can_eliminate_if_unused` method on `Instruction`
https://github.com/noir-lang/noir/blob/ad5a9808ab7c2a89beadb6f95c664441ee27a5f8/compiler/noirc_evaluator/src/ssa/ir/instruction.rs#L534

Reply:
Created https://github.com/noir-lang/noir/issues/7104